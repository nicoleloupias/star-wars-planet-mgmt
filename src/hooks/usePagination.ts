import { useCallback, useMemo, useState } from "react";

export const usePagination = <T>(data?: T[], size: number = 10) => {
  const totalPages = useMemo(() => {
    if (!data) return 0;

    return Math.ceil(data.length / size);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [size, data?.length]);
  const [page, setPage] = useState(0);

  const onNextPage = useCallback(() => {
    setPage((prevState: number) => {
      if (prevState < totalPages) return prevState + 1;
      return prevState;
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [totalPages]);

  const onPrevPage = useCallback(() => {
    setPage((prevState: number) => {
      if (prevState > 0) return prevState - 1;
      return prevState;
    });
  }, []);

  const slicedData = useMemo(() => {
    if (!data) return data;
    return data.slice(page * size, page * size + size);
  }, [data, size, page]);
  const hasNextPage = useMemo(
    () => !(page === totalPages - 1),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [data, page]
  );
  const hasPrevPage = useMemo(
    () => !(page <= 0),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [data, page]
  );

  return {
    data: slicedData,
    page,
    totalPages,
    setPage,
    onPrevPage,
    onNextPage,
    hasNextPage,
    hasPrevPage
  };
};
