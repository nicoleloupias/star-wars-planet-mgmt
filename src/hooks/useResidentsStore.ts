import { create } from "zustand";
import type { Planet } from "./usePlanetsStore";
import type { Resident } from "../services/types/Resident";

interface ResidentsStore {
  residents: Record<Planet["id"], Resident[]>;
  setResident: (planetId: Planet["id"], resident: Resident) => void;
}

export const useResidentsStore = create<ResidentsStore>((set) => ({
  residents: {},
  setResident: (planetId, resident) => {
    return set((state) => {
      return {
        ...state,
        residents: {
          ...state.residents,
          [planetId]: [...(state.residents?.[planetId] || []), resident]
        }
      };
    });
  }
}));
