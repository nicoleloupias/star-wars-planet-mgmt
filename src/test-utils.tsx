import { ChakraProvider } from "@chakra-ui/react";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import React from "react";
import { Router, RouterProvider } from "react-router-dom";
import { theme } from "./theme";
const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <ChakraProvider theme={theme}>
      <Router location={history.location} navigator={history}>
        {children}
      </Router>
    </ChakraProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {}
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => <AllProviders history={history}>{children}</AllProviders>,
    ...options
  });

export { customRender as render };
