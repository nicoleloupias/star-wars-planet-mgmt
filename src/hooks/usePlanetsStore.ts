import { create } from "zustand";
import { Planet } from "../services/types/Planet";

interface PlanetStore {
  planets: Planet[];
  setPlanets: (planets: Planet[]) => void;
}

export const usePlanetsStore = create<PlanetStore>((set, get) => ({
  planets: [],
  setPlanets: (planets) => {
    return set((state) => {
      return { ...state, planets: [...state.planets, ...planets] };
    });
  }
}));
