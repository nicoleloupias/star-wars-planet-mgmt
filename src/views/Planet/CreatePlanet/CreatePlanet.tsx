import type { PlanetFormFieldValues } from "../PlanetForm";
import { PlanetForm } from "../PlanetForm";
import { FormProvider, useForm } from "react-hook-form";
import { Center, Container, Heading, useToast } from "@chakra-ui/react";
import type { Planet } from "../../../hooks/usePlanetsStore";
import { usePlanetsStore } from "../../../hooks/usePlanetsStore";
import { useNavigate } from "react-router-dom";

export const CreatePlanet = () => {
  const form = useForm<PlanetFormFieldValues>({ mode: "all" });
  const addPlanet = usePlanetsStore((state) => state.addPlanet);
  const navigate = useNavigate();
  const toast = useToast();

  const randomIntFromInterval = (min: number, max: number) => {
    return Math.floor(Math.random() * (max - min + 1) + min);
  };

  const createPlanet = (values: PlanetFormFieldValues) => {
    const generatedId = randomIntFromInterval(40, 100).toString();
    const body: Planet = {
      name: values.name,
      diameter: values?.diameter ? values?.diameter.toString() : "unknown",
      population: values?.population ? values?.population.toString() : "unknown",
      climate: values.climate,
      terrain: values.terrain,
      id: generatedId,
      url: `/planets/${generatedId}/`,
      residents: []
    };
    addPlanet(body);
    navigate("/");
    toast({
      status: "success",
      title: `Success! Planet ${values.name} was added to our system.`,
      position: "top-right"
    });
  };

  return (
    <Center flexDir={"column"} py={16}>
      <Container maxW="500px">
        <Heading mb={6}>Create planet</Heading>
        <FormProvider {...form}>
          <PlanetForm onSubmit={form.handleSubmit(createPlanet)} confirmText="Create" />
        </FormProvider>
      </Container>
    </Center>
  );
};
