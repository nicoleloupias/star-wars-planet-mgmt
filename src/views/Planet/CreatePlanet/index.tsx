import { PlanetForm, PlanetFormFieldValues } from "../PlanetForm";
import { FormProvider, useForm } from "react-hook-form";
import { Center, Container, Heading, useToast } from "@chakra-ui/react";
import { Planet, usePlanetsStore } from "../../../hooks/usePlanetsStore";
import { v4 as uuidv4 } from "uuid";
import { useNavigate } from "react-router-dom";

export const CreatePlanet = () => {
  const form = useForm<PlanetFormFieldValues>({ mode: "all" });
  const addPlanet = usePlanetsStore((state) => state.addPlanet);
  const navigate = useNavigate();
  const toast = useToast();

  const randomIntFromInterval = (min: number, max: number) => {
    return Math.floor(Math.random() * (max - min + 1) + min);
  };

  const createPlanet = (values: PlanetFormFieldValues) => {
    const generatedId = randomIntFromInterval(40, 100).toString();
    const body: Planet = {
      name: values.name,
      diameter: values.diameter.toString(),
      population: values.population.toString(),
      climate: values.climate,
      terrain: values.terrain,
      id: generatedId,
      url: `/planets/${generatedId}/`,
      residents: []
    };
    addPlanet(body);
    navigate("/");
    toast({
      status: "success",
      title: "Success! A planet was added to our system.",
      position: "top-right"
    });
  };

  return (
    <Center flexDir={"column"} py={16}>
      <Container maxW="500px">
        <Heading mb={6}>Create planet</Heading>
        <FormProvider {...form}>
          <PlanetForm onSubmit={form.handleSubmit(createPlanet)} confirmText="Create" />
        </FormProvider>
      </Container>
    </Center>
  );
};
