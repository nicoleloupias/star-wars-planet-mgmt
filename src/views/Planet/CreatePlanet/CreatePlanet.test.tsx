import { describe, expect, test } from "vitest";
import { render } from "../../../test-utils";
import type { RenderResult } from "@testing-library/react";
import { act, fireEvent, waitFor } from "@testing-library/react";
import { CreatePlanet } from "./CreatePlanet";
import { usePlanetsStore } from "../../../hooks/usePlanetsStore";
import { createMemoryHistory } from "history";

describe("<CreatePlanet/>", () => {
  const history = createMemoryHistory({
    initialEntries: ["/create"]
  });
  const factoryComponent = () => {
    return render(<CreatePlanet />, { history });
  };

  const fillForm = (component: RenderResult) => {
    act(() => {
      fireEvent.change(component.getByRole("textbox", { name: "Name" }), { target: { value: "Cool planet" } });
      fireEvent.change(component.getByRole("textbox", { name: "Climate" }), { target: { value: "Hot" } });
      fireEvent.change(component.getByRole("spinbutton", { name: "Habitants" }), { target: { value: "1000" } });
      fireEvent.change(component.getByRole("textbox", { name: "Terrain" }), { target: { value: "Terrain" } });
      fireEvent.change(component.getByRole("spinbutton", { name: "Diameter (km)" }), { target: { value: "150" } });
    });
  };

  test("should render succesfully", () => {
    const component = factoryComponent();

    expect(component).toMatchSnapshot();
  });

  test("should create planet correctly", async () => {
    const component = factoryComponent();
    fillForm(component);

    await waitFor(() => {
      expect(component.getByRole("button", { name: /Create/ })).toBeEnabled();
    });

    expect(history.location.pathname).toBe("/create");

    fireEvent.click(component.getByRole("button", { name: /Create/ }));

    await waitFor(() => {
      expect(usePlanetsStore.getState().planets?.length).toBe(1);
      expect(usePlanetsStore.getState().planets?.[0].name).toBe("Cool planet");
      expect(history.location.pathname).toBe("/");
      expect(component.getByText("Success! Planet Cool planet was added to our system.")).toBeInTheDocument();
    });
  });
});
