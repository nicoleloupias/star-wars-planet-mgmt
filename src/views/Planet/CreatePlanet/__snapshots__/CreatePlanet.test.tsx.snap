// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CreatePlanet/> > should render succesfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <div
        class="css-c00vb"
      >
        <div
          class="chakra-container css-3n8ypn"
        >
          <h2
            class="chakra-heading css-qw7hye"
          >
            Create planet
          </h2>
          <form
            class="chakra-stack css-g6wlko"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="name"
                id="field-:r0:-label"
              >
                Name
              </label>
              <div
                class="chakra-input__group css-1y0e7gb"
                data-group="true"
              >
                <input
                  class="chakra-input css-1rrwde8"
                  data-testid="name"
                  id="name"
                  name="name"
                  placeholder="Name"
                  value=""
                />
                <div
                  class="chakra-input__right-element css-1lds0jh"
                />
              </div>
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="diameter"
                id="field-:r1:-label"
              >
                Diameter (km)
              </label>
              <div
                class="chakra-input__group css-1y0e7gb"
                data-group="true"
              >
                <input
                  class="chakra-input css-1rrwde8"
                  data-testid="diameter"
                  id="diameter"
                  name="diameter"
                  placeholder="Diameter (km)"
                  type="number"
                  value=""
                />
                <div
                  class="chakra-input__right-element css-1lds0jh"
                />
              </div>
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="population"
                id="field-:r2:-label"
              >
                Population
              </label>
              <div
                class="chakra-input__group css-1y0e7gb"
                data-group="true"
              >
                <input
                  class="chakra-input css-1rrwde8"
                  data-testid="population"
                  id="population"
                  name="population"
                  placeholder="Population"
                  type="number"
                  value=""
                />
                <div
                  class="chakra-input__right-element css-1lds0jh"
                />
              </div>
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="climate"
                id="field-:r3:-label"
              >
                Climate
              </label>
              <div
                class="chakra-input__group css-1y0e7gb"
                data-group="true"
              >
                <input
                  class="chakra-input css-1rrwde8"
                  data-testid="climate"
                  id="climate"
                  name="climate"
                  placeholder="Climate"
                  value=""
                />
                <div
                  class="chakra-input__right-element css-1lds0jh"
                />
              </div>
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="terrain"
                id="field-:r4:-label"
              >
                Terrain
              </label>
              <div
                class="chakra-input__group css-1y0e7gb"
                data-group="true"
              >
                <input
                  class="chakra-input css-1rrwde8"
                  data-testid="terrain"
                  id="terrain"
                  name="terrain"
                  placeholder="Terrain"
                  value=""
                />
                <div
                  class="chakra-input__right-element css-1lds0jh"
                />
              </div>
            </div>
            <div
              class="css-u8chpe"
            >
              <button
                class="chakra-button css-1bvqs64"
                type="button"
              >
                Cancel
              </button>
              <button
                class="chakra-button css-advwry"
                disabled=""
                type="submit"
              >
                Create
              </button>
            </div>
          </form>
        </div>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="css-c00vb"
    >
      <div
        class="chakra-container css-3n8ypn"
      >
        <h2
          class="chakra-heading css-qw7hye"
        >
          Create planet
        </h2>
        <form
          class="chakra-stack css-g6wlko"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-g6pte"
              for="name"
              id="field-:r0:-label"
            >
              Name
            </label>
            <div
              class="chakra-input__group css-1y0e7gb"
              data-group="true"
            >
              <input
                class="chakra-input css-1rrwde8"
                data-testid="name"
                id="name"
                name="name"
                placeholder="Name"
                value=""
              />
              <div
                class="chakra-input__right-element css-1lds0jh"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-g6pte"
              for="diameter"
              id="field-:r1:-label"
            >
              Diameter (km)
            </label>
            <div
              class="chakra-input__group css-1y0e7gb"
              data-group="true"
            >
              <input
                class="chakra-input css-1rrwde8"
                data-testid="diameter"
                id="diameter"
                name="diameter"
                placeholder="Diameter (km)"
                type="number"
                value=""
              />
              <div
                class="chakra-input__right-element css-1lds0jh"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-g6pte"
              for="population"
              id="field-:r2:-label"
            >
              Population
            </label>
            <div
              class="chakra-input__group css-1y0e7gb"
              data-group="true"
            >
              <input
                class="chakra-input css-1rrwde8"
                data-testid="population"
                id="population"
                name="population"
                placeholder="Population"
                type="number"
                value=""
              />
              <div
                class="chakra-input__right-element css-1lds0jh"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-g6pte"
              for="climate"
              id="field-:r3:-label"
            >
              Climate
            </label>
            <div
              class="chakra-input__group css-1y0e7gb"
              data-group="true"
            >
              <input
                class="chakra-input css-1rrwde8"
                data-testid="climate"
                id="climate"
                name="climate"
                placeholder="Climate"
                value=""
              />
              <div
                class="chakra-input__right-element css-1lds0jh"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-g6pte"
              for="terrain"
              id="field-:r4:-label"
            >
              Terrain
            </label>
            <div
              class="chakra-input__group css-1y0e7gb"
              data-group="true"
            >
              <input
                class="chakra-input css-1rrwde8"
                data-testid="terrain"
                id="terrain"
                name="terrain"
                placeholder="Terrain"
                value=""
              />
              <div
                class="chakra-input__right-element css-1lds0jh"
              />
            </div>
          </div>
          <div
            class="css-u8chpe"
          >
            <button
              class="chakra-button css-1bvqs64"
              type="button"
            >
              Cancel
            </button>
            <button
              class="chakra-button css-advwry"
              disabled=""
              type="submit"
            >
              Create
            </button>
          </div>
        </form>
      </div>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
